### PR TITLE
bug(techdocs): Internal techdocs anchor elements do not navigate correctly when Backstage is on a subpath

### PR DESCRIPTION
**Pull request by sennyeya**
## Hey, I just made a Pull Request!

Fixes https://github.com/backstage/backstage/issues/16544. 
Basically, the `react-router-dom`'s `navigate` method only accepts paths (without the host, origin, or anything else). We were passing it the current page's path without the possible `app.baseUrl` sub-path removed. This work removes that sub-path when present and provides a wrapper hook that exposes this functionality for possible re-use.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/core-app-api/src/routing/useNavigateUrl.tsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ *
+ * NOTE: This does not support routing to external URLs. That should be done with a `Link` or `a`
+ *  element instead, or just `window.location.href`.
+ *
+ * @returns Navigation function that is a wrapper over `react-router-dom`'s
+ *  to support passing full URLs for navigation.
+ *
+ * @public
+ */
+export function useNavigateUrl() {
```
Comment (Rugvip) [inline packages/core-app-api/src/routing/useNavigateUrl.tsx:54]: A lot of the functionality baked in here is already implemented by the routing system. We've discussed this need before and what I suggested back then was that we add a `appRootRouteRef` that references the root of the app. I think that's the leanest way to go about this that ships this need with the fewest API additions. From what I can tell it'll work to solve this use-case. We'd export that from `@backstage/core-plugin-api`.

What do you think?

**Review by Rugvip — CHANGES_REQUESTED** [review]
Nice to see this solved, thanks! 🎉 

Some thoughts on how we expose this through the core APIs, looks good otherwise 👍

File: packages/core-app-api/src/routing/useNavigateUrl.tsx (hunk 1 above)
Comment (sennyeya) [reply packages/core-app-api/src/routing/useNavigateUrl.tsx:54]: I agree that this hook isn't needed and would fit better in the existing routing API but I'm not completely following. I took a look at the `@backstage/core-plugin-api` `routing` folder and it seems relevant but not sure how just adding `appRootRouteRef` would solve this issue. It looks like most of the existing routing is meant to translate a `RouteRef` into its corresponding URL not the other way around.

Comment (sennyeya) [reply packages/core-app-api/src/routing/useNavigateUrl.tsx:54]: @Rugvip wanted to follow up on this, can you explain a little bit more your thinking here?

**Comment by artemptushkin** [comment]
@sennyeya @Rugvip hey, any chance to proceed with this PR? This would help the issue for us (I hope)

File: packages/core-app-api/src/routing/useNavigateUrl.tsx (hunk 1 above)
Comment (Rugvip) [reply packages/core-app-api/src/routing/useNavigateUrl.tsx:54]: My thinking is that you'd be able to do this:

```tsx
import { appRootRouteRef, useRouteRef } from `@backstage/core-plugin-api`;

function MyComponent() {
  const appRootLink = useRouteRef(appRootRouteRef);

  return <a href={appRootLink()}>Link to app root</a>
}
```

The `appRootRouteRef` would simply be bound to `/` in the app, which the routing system would then resolve with the appropriate base URL.

I see what you mean though, as what you need is to be able to remove that `appRootLink()` from an existing URL. Tbh I feel this is too specific of a need to add to the core APIs. The `appRootRouteRef` does take you halfway though, it avoids needing to read `app.baseUrl` from config, which might not always be correct, the existing logic for that in the app is [here](https://github.com/backstage/backstage/blob/b9828c3c8dcdd72329f190cad801cd0388999b34/packages/core-app-api/src/app/AppRouter.tsx#L37-L59). With that in place I think the rest of the logic should be contained within TechDocs for now. Alternatively we skip the `appRootRouteRef` and duplicate the config reading logic in TechDocs too, but it's a lil bit more fragile.

File: plugins/techdocs/package.json [hunk 2]
```diff
@@ -35,6 +35,7 @@
   "dependencies": {
     "@backstage/catalog-model": "workspace:^",
     "@backstage/config": "workspace:^",
+    "@backstage/core-app-api": "workspace:^",
```
Comment (Rugvip) [inline plugins/techdocs/package.json:38]: This needs to stay in dev deps

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you, almost good to go, just a tiny tweak

File: plugins/techdocs/package.json (hunk 2 above)
Comment (sennyeya) [reply plugins/techdocs/package.json]: Updated!

**Review by Rugvip — APPROVED** [review]
Great, thank you! 🎉 

Let's :shipit:

**Comment by Rugvip** [comment]
@backstage/techdocs-maintainers, will need a 🟢 from you too

**Review by agentbellnorm — APPROVED** [review]
Nice work, looks good to me!

**Pull request merged**